### ISSUE TITLE
Flatpak app

### ISSUE THREAD

**Issue by tim77** [theia-ide/theia-apps#341]
Have you considered make universal package which will run on all major distributions? There already some [initial work](https://github.com/gasinvein/flatpak-apps-staging/tree/org.theia_ide.theia-electron/master) done (thanks to @gasinvein) and it compiles and works.

- [Flatpak website](https://www.flatpak.org/)
- [Flatpak docs](http://docs.flatpak.org/en/latest/getting-started.html)

**Comment by vince-fugnitto** [comment]
@tim77 I don't believe we were aware of Flatpack so thank you for sharing.

For our electron applications, we use electron-builder which has the option of creating [AppImages](https://appimage.org/) which ultimately satisfy the requirements of being able to run on all major Linux distributions. ElectronBuilder also allows us to create packages for Mac and Windows as well. 

I'm not sure theres a need to have another type of packaging in the form of Flatpak.

**Comment by tim77** [comment]
> Would flatpak be better, as far as you know?

Flatpak much more portable and guaranty to work on old and new distros. Also your app will be available out of box in modern GUI package managers like GNOME Software and KDE Discover. This is attractive for users and help to discover your app.

I'll add more details about this soon.

**Comment by vince-fugnitto** [comment]
I found the following comparison resource for reference (might be biased since it comes from an appimage repo):
- https://github.com/AppImage/AppImageKit/wiki/Similar-projects#comparison

**Comment by gasinvein** [comment]
@marcdumais-work 
>Would flatpak be better, as far as you know?

In terms of ABI portability - yes, since flatpak app is completely isolated from host, it's kinda more like Docker than AppImage.

**Comment by tim77** [comment]
There is some interesting and informative talks [here](https://github.com/obsproject/rfcs/pull/21). Hope you find this interesting.

**Comment by vince-fugnitto** [comment]
> There is some interesting and informative talks [here](https://github.com/obsproject/rfcs/pull/21). Hope you find this interesting.

Thank you for the information! 
In reality, application developers of Theia-based IDEs are free to use any packaging and distribution tools they wish, and this repo is mainly used for examples, and test purposes. If you believe that a working example would be useful to include in the repository please feel free to submit a pull-request. It can be used by others for reference who may have the same goal :)

**Comment by sudo-give-me-coffee** [comment]
> e.g. would it be possible to build on Ubuntu 18.04 and then run on Red Hat 7?

Yes, Electron apps runs even on Ubuntu 14.04 (a very older OS than thay Flatpak supports)

**Comment by sudo-give-me-coffee** [comment]
> Flatpak much more portable and guaranty to work on old and new distros. Also your app will be available out of box in modern GUI package managers like GNOME Software and KDE Discover. 

Only if user has added support for Flatpak and repository that application was published

**Comment by tim77** [comment]
> > e.g. would it be possible to build on Ubuntu 18.04 and then run on Red Hat 7?
> 
> Yes, Electron apps runs even on Ubuntu 14.04 (a very older OS than thay Flatpak supports)

Uh, this goes against with what says [official](https://docs.appimage.org/reference/best-practices.html#general-recommendations) AppImages docs and this not even joke, as already said. 😄
![DeepinScreenshot_select-area_20200427232303](https://user-images.githubusercontent.com/5614476/80472763-d6f3a400-894d-11ea-8d23-68f6499da5a7.png)

@vince-fugnitto i'll ping @gasinvein since it wrote already working manifest, thank you.

Also, linking with OBS issue was really bad idea, sorry. I forgot that this will be visible on linked issue and all what i wanted is to share some GeorgesStavracas points and information. Please lock this thread since there productive and useful conversation not possible because of this.

**Comment by sudo-give-me-coffee** [comment]
@tim77 Just open the link, they explain the reason for building on older distros :wink:

And Electron has tingly difference of traditional C or C++, since they was builded on top of Ubuntu 12.04 any distro with glibc 2.12 will run electron apps.  Yesterday, i was  tested with TinyCore and worked fine (app created on Arch Linux up to date), flatpak will be good for supporting Alpine Linux and other libmusl based distros

**Comment by marcdumais-work** [comment]
> i wanted is to share some GeorgesStavracas points and information

Thanks @tim77 - interesting read. No worries about linking. I think we'd like to have an example of using `Flatpack` for package a `Theia` app. Do you prefer we open a new issue about that and leave this one here closed? Else I would just re-open it. 

Generally, in this context of this repo here, we have the luxury of not having to pick and choose. Any given __`Theia`-based product__ will need to decide what's best wr to their own needs and according to the detailed strenghts and weaknesses of each potential packaging. We are happy to by-pass the debate and offer both side-by-side :)

ATM our `theia-electron` example application uses `Electron Builder` which offers `AppImage` as one of its output formats on Linux hosts. Maybe it will eventually also support `Flatpack`? In the meantime we'll certainly consider having a standalone Flatpack example in this repo and would be open to have it contributed. Else we'll get to it eventually.

**Comment by tim77** [comment]
@marcdumais-work 
> Maybe it will eventually also support Flatpack?

The [request](https://github.com/electron-userland/electron-builder/issues/3212) for support Flatpak in Electron Builder exist for a long time but seems like stalled.

Just look [how it could look](https://flathub.org/apps/details/com.visualstudio.code) handy from user perspective when they have central place for Apps which is safe to use and easy to install/update.

**Comment by vince-fugnitto** [comment]
> @marcdumais-work
> 
> > Maybe it will eventually also support Flatpack?
> 
> The [request](https://github.com/electron-userland/electron-builder/issues/3212) for support Flatpak in Electron Builder exist for a long time but seems like stalled.

I'm curious to know why electron-builder has not supported flatpak yet, perhaps it does not see the value in supporting yet another type of packaging?

> Just look [how it could look](https://flathub.org/apps/details/com.visualstudio.code) handy from user perspective when they have central place for Apps which is safe to use and easy to install/update.

I guess it's the same outcome of doing something like #280 and publishing to the official Electron marketplace?

**Comment by marcdumais-work** [comment]
> I guess it's the same outcome of doing something like #280 and publishing to the official Electron marketplace?

Sure. Let's see which one gets contributed here first :1st_place_medal:  :)

**Comment by tim77** [comment]
@gasinvein updated Theia app in his hub, i asked him send a pull but we cannot find agreement there yet. He said that we should properly made fully automated CI build but this is not even possible at this stage since Theia is a new app and doesn't have fixed releases. Modules using `next` version so every build will be non-reproducible. But in same time IMO something is better then nothing. It compiles and works fine. Note: for `generated-sources.json` you need slightly [patched](https://github.com/refi64/flatpak-builder-tools/pull/1) helper which will be upstreamed soon i hope.

> I guess it's the same outcome of doing something like #280 and publishing to the official Electron marketplace?

Yes, but with one remark: Flatpak mean you packaged your app for Linux users, Snap mean you packaged it for Ubuntu. There is some article about [adoption](https://kamikazow.wordpress.com/2018/06/08/adoption-of-flatpak-vs-snap-2018-edition/), it may look outdated, but nothing really changed since then. Devs wanted to [drop](https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/253) completely Snap support in GNOME Software but postponed this change.

**Comment by marcdumais-work** [comment]
> He said that we should properly made fully automated CI build but this is not even possible at this stage since Theia is a new app and doesn't have fixed releases. Modules using `next` version so every build will be non-reproducible.

For most apps we have under CI in this repo, we have both a `latest` based on the latest monthly Theia release and a `next` based on latest Theia master. I do not see why we could not publish in Flatpack format the same or similar Theia app as [theia-docker](https://hub.docker.com/r/theiaide/theia). It could be only `latest` if there's a concern about potential random issues with `next`.

**Comment by deknos** [comment]
just my 2 cent: as long as there's no appimage repository hub which is recognized within the community, from which i can pull update, i would prefer flatpak, where i can pull updates from the developers or from flathub or sth like that. appimages on my system tend to never get updated.
